src/tiledb/cloud/vcf/vcf_toolbox/annotate.py

```diff
@@ -0,0 +1,227 @@
+import logging
+import time
+from concurrent.futures import ThreadPoolExecutor
+from typing import Mapping, Optional, Sequence, Tuple, Union
+
+import pandas as pd
+
+import tiledb
+from tiledb.cloud.utilities import get_logger
+from tiledb.cloud.utilities import max_memory_usage
+
+from .transform import df_transform
+
+
+def split_region(region: str) -> Tuple[str, slice]:
+    """
+    Split a region string into contig and slice.
+
+    The format of the region string is 'contig:start-end'.
+
+    :param region: region to split
+    :raises ValueError: if region is not in the format 'contig:start-end'
+    :return: contig and slice
+    """
+
+    if not isinstance(region, str) or ":" not in region or "-" not in region:
+        raise ValueError("Region must be in the format 'contig:start-end'")
+    contig, range = region.split(":")
+    start, end = range.split("-")
+    return contig, slice(int(start), int(end))
+
+
+def split_regions(regions: Union[str, Sequence[str]]) -> Tuple[str, Sequence[slice]]:
+    """
+    Split a region or list of regions into a contig and list of slices.
+
+    :param regions: regions to split
+    :raises ValueError: if regions are not in the same contig
+    :return: contig and list of slices
+    """
+
+    if isinstance(regions, str):
+        regions = [regions]
+
+    contig = None
+    ranges = []
+    for region in regions:
+        c, r = split_region(region)
+        if contig is None:
+            contig = c
+        elif contig != c:
+            raise ValueError("All regions must be in the same contig")
+        ranges.append(r)
+
+    return contig, ranges
+
+
+def zygosity(gt: Tuple[int, int]) -> str:
+    """
+    Convert genotype to a zygosity string.
+
+    :param gt: genotype
+    :return: zygosity string
+    """
+    if tuple(gt) == tuple([0, 0]):
+        return "HOM_REF"
+    if tuple(gt) == tuple([".", "."]):
+        return "UNKNOWN"
+    elif len(tuple(gt)) == 2 and (tuple(gt)[0] == 0 or tuple(gt)[1] == 0):
+        return "HET"
+    elif tuple(gt)[0] != 0 and tuple(gt)[1] != 0:
+        return "HOM_ALT"
+    elif len(tuple(gt)) == 1 and tuple(gt)[0] != 0:
+        return "HEMI"
+    else:
+        return str(gt)
+
+
+def _annotate(
+    vcf_df: pd.DataFrame,
+    *,
+    ann_uri: str,
+    ann_regions: Union[str, Sequence[str]],
+    ann_attrs: Optional[Union[str, Sequence[str]]] = None,
+    vcf_filter: Optional[str] = None,
+    split_multiallelic: bool = True,
+    add_zygosity: bool = False,
+    reorder: Optional[Sequence[str]] = [
```

Comment:
Thinking about it, the `reorder` and `rename` args can be `None` by default since they are mainly cosmetic. We can leave the data in the default format (column names and column order) and let the data consumer decide how they want the data presented.